nomad/structs/csi.go

```diff
@@ -874,6 +890,128 @@ func (p *CSIPlugin) DeleteAlloc(allocID, nodeID string) error {
 	return nil
 }
 
+// AddJob adds a job to the plugin and increments expected
+func (p *CSIPlugin) AddJob(job *Job, summary *JobSummary) {
+	p.UpdateExpectedWithJob(job, summary, false)
+}
+
+// DeleteJob removes the job from the plugin and decrements expected
+func (p *CSIPlugin) DeleteJob(job *Job, summary *JobSummary) {
+	p.UpdateExpectedWithJob(job, summary, true)
+}
+
+// UpdateExpectedWithJob maintains the expected instance count
+// we use the summary to add non-allocation expected counts
+func (p *CSIPlugin) UpdateExpectedWithJob(job *Job, summary *JobSummary, terminal bool) {
+	var count int
+
+	for _, tg := range job.TaskGroups {
+		if job.Type == JobTypeSystem {
+			if summary == nil {
+				continue
+			}
+
+			s, ok := summary.Summary[tg.Name]
+			if !ok {
+				continue
+			}
+
+			count = s.Running + s.Queued + s.Starting
+		} else {
+			count = tg.Count
+		}
+
+		for _, t := range tg.Tasks {
+			if t.CSIPluginConfig == nil ||
+				t.CSIPluginConfig.ID != p.ID {
+				continue
+			}
+
+			// Change the correct plugin expected, monolith should change both
+			if t.CSIPluginConfig.Type == CSIPluginTypeController ||
+				t.CSIPluginConfig.Type == CSIPluginTypeMonolith {
+				if terminal {
+					p.ControllerJobs.Delete(job)
+				} else {
+					p.ControllerJobs.Add(job, count)
+				}
+			}
+
+			if t.CSIPluginConfig.Type == CSIPluginTypeNode ||
+				t.CSIPluginConfig.Type == CSIPluginTypeMonolith {
+				if terminal {
+					p.NodeJobs.Delete(job)
+				} else {
+					p.NodeJobs.Add(job, count)
+				}
+			}
+		}
+	}
+
+	p.ControllersExpected = p.ControllerJobs.Count()
+	p.NodesExpected = p.NodeJobs.Count()
+}
+
+// JobDescription records Job identification and the count of expected plugin instances
+type JobDescription struct {
+	Namespace string
+	ID        string
+	Version   uint64
+	Expected  int
+}
+
+// JobNamespacedDescriptions maps Job.ID to JobDescription
+type JobNamespacedDescriptions map[string]JobDescription
+
+// JobDescriptions maps Namespace to a mapping of Job.ID to JobDescription
+type JobDescriptions map[string]JobNamespacedDescriptions
+
+func (j JobDescriptions) Add(job *Job, expected int) {
+	if j == nil {
+		j = make(JobDescriptions)
+	}
+	if j[job.Namespace] == nil {
+		j[job.Namespace] = make(JobNamespacedDescriptions)
+	}
+	j[job.Namespace][job.ID] = JobDescription{
+		Namespace: job.Namespace,
+		ID:        job.ID,
+		Version:   job.Version,
```

Comment:
If we're using the version as a value and not part of the key, that implies to me that the later version of a job is intended to overwrite the values for a newer version. Can you walk me through what we expect to see here in the case of a job update with canaries or a rollback?

Reply:
I've talked to @DingoEatingFuzz, and the API changes are not necessary; the UI has to make the same number of round trips to get the required data. I think for now the best idea is just to back out the API change, which I've done.